### PR TITLE
Also use NotifyDirectCallTaskBlock/Unblocked for plasma store accesses

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

Direct call tasks must set blocked/unblocked via its own IPC to the raylet. In a previous change, we did this for memory store objects, but not for plasma objects.

### CODE REVIEW

File: src/ray/core_worker/store_provider/plasma_store_provider.cc
Comment (ericl) [inline src/ray/core_worker/store_provider/plasma_store_provider.cc]: @edoakes it looks to me Wait() wasn't block / unblocking. Was this a bug in the core worker implementation of Wait?

File: src/ray/core_worker/store_provider/plasma_store_provider.cc
Comment (stephanie-wang) [inline src/ray/core_worker/store_provider/plasma_store_provider.cc]: Can you add a comment explaining under what cases we reach this line?

File: src/ray/core_worker/store_provider/plasma_store_provider.cc
Comment (stephanie-wang) [inline src/ray/core_worker/store_provider/plasma_store_provider.cc]: I think we could avoid making the changes in node_manager.cc if we just set the `fetch_only` flag correctly: something like `in_direct_call ? true : fetch_only`. Maybe less explicit but this was also a bit confusing to me.

File: src/ray/core_worker/store_provider/plasma_store_provider.cc
Comment (stephanie-wang) [inline src/ray/core_worker/store_provider/plasma_store_provider.cc]: How about we do the same thing as for the raylet path and call `HandleDirectCallTaskBlocked/Unblocked` directly from the node manager when processing/returning from the `Wait`? I think that's cleaner than doing separate IPCs from the client.

**Review by stephanie-wang — COMMENTED** [review]
Eventually, it'd be nice if we could only call block/unblock on the higher-level `CoreWorker::Get`, but it does seem pretty annoying to do...

File: src/ray/core_worker/store_provider/plasma_store_provider.cc
Comment (stephanie-wang) [inline src/ray/core_worker/store_provider/plasma_store_provider.cc]: We could use the in_direct_call flag to directly call `HandleDirectCallTaskBlocked` from the node_manager, in the `ProcessFetchOrReconstructMessage`.

File: src/ray/core_worker/store_provider/plasma_store_provider.cc
Comment (ericl) [reply src/ray/core_worker/store_provider/plasma_store_provider.cc]: Fixing.

File: src/ray/core_worker/store_provider/plasma_store_provider.cc
Comment (ericl) [reply src/ray/core_worker/store_provider/plasma_store_provider.cc]: I'm not sure I follow. Isn't fetch only separate?

File: src/ray/core_worker/store_provider/plasma_store_provider.cc
Comment (ericl) [reply src/ray/core_worker/store_provider/plasma_store_provider.cc]: The reason I did it this way is for blocking and unblocking to be consistently done in the new path. It's confusing when sometimes blocking is implicit and sometimes not. Performance wise, we are in the slow path for direct calls so an extra IPC can't hurt.

File: src/ray/core_worker/store_provider/plasma_store_provider.cc
Comment (ericl) [reply src/ray/core_worker/store_provider/plasma_store_provider.cc]: The reason that doesn't work is that we need to check if we are also in the main thread and not in an actor method. We could pass this metadata all the way to the raylet, but it is simpler to decide here and send the IPC separately.

File: src/ray/core_worker/store_provider/plasma_store_provider.cc
Comment (zhijunfu) [inline src/ray/core_worker/store_provider/plasma_store_provider.cc]: Just wondering is it required for `NotifyDirectCallTaskBlocked` to be called before `Wait`?  

Would it work if we reverse the order of the two operations? I'm not suggest this change, but just curious.

Comment (ericl) [reply src/ray/core_worker/store_provider/plasma_store_provider.cc]: Wait is a blocking call, so I think it would basically mean the block call would become a no-op if done afterwards.

File: src/ray/core_worker/store_provider/plasma_store_provider.cc
Comment (stephanie-wang) [reply src/ray/core_worker/store_provider/plasma_store_provider.cc]: The new flag seems to do exactly the same thing as fetch_only=true, so I was suggesting we just reuse the fetch_only flag. So instead of passing both fetch_only and mark_worker_blocked as fields into FetchOrReconstruct, we'd just pass `in_direct_call ? true : fetch_only`.

File: src/ray/core_worker/store_provider/plasma_store_provider.cc
Comment (stephanie-wang) [reply src/ray/core_worker/store_provider/plasma_store_provider.cc]: Okay, sounds good.

**Review by stephanie-wang — COMMENTED** [review]
Hmm I just realized, isn't there an issue with the HandleDirectCallTaskBlocked/Unblocked handlers if they get processed out-of-order? Although, I guess the worst case is that we'll think there's more CPU available than there actually is, so maybe that's fine...

File: src/ray/core_worker/store_provider/plasma_store_provider.cc
Comment (zhijunfu) [reply src/ray/core_worker/store_provider/plasma_store_provider.cc]: I was just thinking of the out of order issue of grpc - it's fine for now since we're still using raw tcp connection to communicate with raylet, but if later we want to migrate to grpc then this could be an issue. I saw @stephanie-wang had a similar comment.

**Comment by ericl** [comment]
> Hmm I just realized, isn't there an issue with the HandleDirectCallTaskBlocked/Unblocked handlers if they get processed out-of-order? Although, I guess the worst case is that we'll think there's more CPU available than there actually is, so maybe that's fine...

It's currently guaranteed to be in order. There could be a race with returning the task lease, which I handled by having the task lease call force unblock the worker when it gets returned (unblock is idempotent).

File: src/ray/core_worker/store_provider/plasma_store_provider.cc
Comment (ericl) [reply src/ray/core_worker/store_provider/plasma_store_provider.cc]: After looking into this flag more I'm not even sure this PR is correct currently now, let's follow up offline.

File: src/ray/core_worker/store_provider/plasma_store_provider.cc
Comment (stephanie-wang) [inline src/ray/core_worker/store_provider/plasma_store_provider.cc]: Hmm actually won't this break the eviction errors? Because right now, we rely on the call to the task_dependency_manager_ [here](https://github.com/ray-project/ray/blob/master/src/ray/raylet/node_manager.cc#L1953) to cause the Get to timeout and store an UnreconstructableError.

Comment (ericl) [reply src/ray/core_worker/store_provider/plasma_store_provider.cc]: Hm I see, so IIUC fetch won't store errors?

That sounds unfortunate, so do we need to restore the mark_worker_blocked flag and pass it all the way into HandleTaskBlocked (maybe rename HandleTaskBlocked to something more sensible?)

Comment (stephanie-wang) [reply src/ray/core_worker/store_provider/plasma_store_provider.cc]: Yeah, sorry I didn't realize this earlier.

Another option is to duplicate the task dependency manager calls in the HandleDirectCallTaskBlocked/Unblocked handlers. It's not clear to me which one would be cleaner (aka less messy).

Comment (ericl) [reply src/ray/core_worker/store_provider/plasma_store_provider.cc]: I ended up doing the former. It is quite messy now, but I guess once we delete the old path we can remove the code guarded by the flag entirely and it will be clean again.

File: src/ray/raylet/node_manager.cc
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.cc]: I think we need to do this for HandleDirectCallTaskUnblocked as well...

**Review by stephanie-wang — COMMENTED** [review]
Hmm I think there is still an issue with unblocking the task. It looks like this will keep the Get call open in the raylet for direct tasks even if the objects are local and the task is no longer blocked (see comment in the code).

File: src/ray/raylet/node_manager.h
Comment (ericl) [reply src/ray/raylet/node_manager.h]: Done

File: src/ray/raylet/node_manager.cc
Comment (ericl) [reply src/ray/raylet/node_manager.cc]: Done

**Review by stephanie-wang — APPROVED** [review]

File: src/ray/core_worker/context.cc
Comment (edoakes) [inline src/ray/core_worker/context.cc]: comment explaining this condition would be helpful

File: src/ray/core_worker/store_provider/memory_store/memory_store.cc
Comment (edoakes) [inline src/ray/core_worker/store_provider/memory_store/memory_store.cc]: Is the `raylet_client_` actually `nullptr` under non-error conditions?

File: src/ray/core_worker/store_provider/plasma_store_provider.cc
Comment (edoakes) [inline src/ray/core_worker/store_provider/plasma_store_provider.cc]: `is_direct_call`

File: src/ray/raylet/format/node_manager.fbs
Comment (edoakes) [inline src/ray/raylet/format/node_manager.fbs]: Is this comment still correct?

File: src/ray/raylet/format/node_manager.fbs
Comment (edoakes) [inline src/ray/raylet/format/node_manager.fbs]: Same as above.

File: src/ray/raylet/format/node_manager.fbs
Comment (edoakes) [inline src/ray/raylet/format/node_manager.fbs]: More info about how this is actually handled would be more helpful

Comment (ericl) [reply src/ray/raylet/format/node_manager.fbs]: Ah right, clarified.

File: src/ray/raylet/format/node_manager.fbs
Comment (ericl) [reply src/ray/raylet/format/node_manager.fbs]: Removed the comment.

File: src/ray/core_worker/store_provider/memory_store/memory_store.cc
Comment (ericl) [reply src/ray/core_worker/store_provider/memory_store/memory_store.cc]: Yea see the header.

**Comment by ericl** [comment]
Consistently /python/ray/tests:test_failure                                         TIMEOUT in 315.3s

**Comment by stephanie-wang** [comment]
> Consistently /python/ray/tests:test_failure TIMEOUT in 315.3s

If it's one of the direct call tests, then that sounds like it's because an eviction error isn't getting thrown properly. Probably because task_dependency_manager_.SubscribeGetDependencies isn't being called somewhere.

**Comment by ericl** [comment]
Yeah that was it, hopefully fixed now.

**Comment by ericl** [comment]
Hmm the matrix test passes after merging with master, let's try it again.

**Pull request merged**
